README.md
<h1 align="center">vsd</h1>

<p align="center">
  <a href="https://github.com/clitic/vsd">
    <img src="https://img.shields.io/github/downloads/clitic/vsd/total?logo=github&style=flat-square">
  </a>
  <a href="https://crates.io/crates/vsd">
    <img src="https://img.shields.io/crates/d/vsd?logo=rust&style=flat-square">
  </a>
  <a href="https://crates.io/crates/vsd">
    <img src="https://img.shields.io/crates/v/vsd?style=flat-square">
  </a>
  <a href="https://github.com/clitic/vsd#license">
    <img src="https://img.shields.io/crates/l/vsd?style=flat-square">
  </a>
  <a href="https://github.com/clitic/vsd">
    <img src="https://img.shields.io/github/repo-size/clitic/vsd?logo=github&style=flat-square">
  </a>
  <a href="https://github.com/clitic/vsd">
    <img src="https://img.shields.io/tokei/lines/github/clitic/vsd?style=flat-square">
  </a>
  <a href="https://colab.research.google.com/github/clitic/vsd/blob/main/vsd-on-colab.ipynb">
    <img src="https://img.shields.io/badge/Open%20In%20Colab-F9AB00?logo=googlecolab&color=525252&style=flat-square">
  </a>
</p>

<p align="center">
  <a href="#Installation">Installation</a>
  &nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;
  <a href="#Usage">Usage</a>
  &nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;
  <a href="https://colab.research.google.com/github/clitic/vsd/blob/main/vsd-on-colab.ipynb">Try Without Install</a>
</p>

**v**ideo **s**tream **d**ownloader is a command line program to download video streams served over HTTP from websites, [HLS](https://howvideo.works/#hls) and [DASH](https://howvideo.works/#dash) playlists.

<p align="center">
  <img src="https://github.com/clitic/vsd/blob/main/images/showcase.gif">
</p>

## Features

- [x] Capturing network requests and collecting .m3u8, .mpd and subtitles from websites and save them locally.
- [x] Muxing streams to single video container using ffmpeg.
- [x] Singular progress bar for complete download process like an normal file download with realtime file size estimations.
- [x] Supports `AES-128` and `CENC` playlists decryption.
- [x] Supports HLS and DASH
- [x] Supports downloading in multiple threads.
- [ ] GUI (maybe in future)
- [ ] Supports [SAMPLE-AES](https://developer.apple.com/library/archive/documentation/AudioVideo/Conceptual/HLS_Sample_Encryption/Encryption/Encryption.html) playlist decryption.
- [ ] Live stream download (wip)

<a href="#Help">See More</a>

## Installation
  
Dependencies

- [ffmpeg](https://www.ffmpeg.org/download.html) (optional, *recommended*) only required for transmuxing and transcoding streams.
- [chrome](https://www.google.com/chrome) / [chromium](https://www.chromium.org/getting-involved/download-chromium/) (optional) only required for `capture` and `collect` subcommands. 

Visit [releases](https://github.com/clitic/vsd/releases) for prebuilt binaries. Download and extract archive and copy vsd binary to any path. Now add that path to your `PATH` environment variable. Build instructions can be found [here](https://github.com/clitic/vsd/blob/main/BUILD.md) and changelog [here](https://github.com/clitic/vsd/blob/main/CHANGELOG.md).

| Host                | Architecture | Download                                                                                                      | Install                                                                                |
|---------------------|--------------|---------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------|
| Android 7+ (Termux) | aarch64      | [.tar.gz](https://github.com/clitic/vsd/releases/download/v0.2.5/vsd-v0.2.5-aarch64-linux-android.tar.gz)     | [command](https://github.com/clitic/vsd/blob/main/INSTALL.md#android-7-termux-aarch64) |
| Linux               | x86_64       | [.tar.gz](https://github.com/clitic/vsd/releases/download/v0.2.5/vsd-v0.2.5-x86_64-unknown-linux-musl.tar.gz) | [command](https://github.com/clitic/vsd/blob/main/INSTALL.md#linux-x86_64)             |
| MacOS 12.3+         | x86_64       | [.tar.gz](https://github.com/clitic/vsd/releases/download/v0.2.5/vsd-v0.2.5-x86_64-apple-darwin.tar.gz)       | [command](https://github.com/clitic/vsd/blob/main/INSTALL.md#macos-123-x86_64)         |
| Windows             | x86_64       | [.zip](https://github.com/clitic/vsd/releases/download/v0.2.5/vsd-v0.2.5-x86_64-pc-windows-msvc.zip)          |                                                                                        |

## Usage

For quick testing purposes you may use [https://test-streams.mux.dev](https://test-streams.mux.dev) as direct input. These streams are used by [hls.js](https://github.com/video-dev/hls.js) for testing purposes.

- Downloading and saving HLS and DASH playlists to disk.

```bash
$ vsd save <url> -o video.mp4
```

- Collecting .m3u8 (HLS), .mpd (Dash) and subtitles from a website and saving them locally.

```bash
$ vsd collect <url>
```

## Help

```bash
$ vsd --help
```

```
Download video streams served over HTTP from websites, HLS and DASH playlists

Usage: vsd.exe <COMMAND>

Commands:
  capture  Capture requests made to fetch playlists
  collect  Collect playlists and subtitles from a website and save them locally
  decrypt  Decrypt encrypted streams using keys
  extract  Extract subtitles embedded inside an mp4 file
  merge    Merge multiple segments to a single file
  save     Download and save HLS and DASH playlists to disk
  help     Print this message or the help of the given subcommand(s)

Options:
  -h, --help     Print help information (use `--help` for more detail)
  -V, --version  Print version information
```

```bash
$ vsd save --help
```

```
Download and save HLS and DASH playlists to disk

Usage: vsd.exe save [OPTIONS] <INPUT>

Arguments:
  <INPUT>  http(s):// | .m3u8 | .m3u | .mpd | .xml

Options:
  -a, --alternative
          Download alternative audio or subtitles stream from playlist instead all streams. For
          downloading video stream only, use `--skip` flag
      --baseurl <BASEURL>
          Base url for all segments. Usually needed for local m3u8 file
  -d, --directory <DIRECTORY>
          Change directory path for temporarily downloaded files. By default current working
          directory is used
  -k, --key <<KID:(base64:)KEY>|(base64:)KEY>
          Decryption keys for decrypting CENC encrypted streams. Key value should be specified in
          hex. Use `base64:` prefix if key is in base64 format. Streams encrypted with a single key
          can use `--key base64:MhbcGzyxPfkOsp3FS8qPyA==` like key format. Streams encrypted with
          multiple keys can use `--key
          eb676abbcb345e96bbcf616630f1a3da:100b6c20940f779a4589152b57d2dacb like key format. This
          option can be used multiple times
  -o, --output <OUTPUT>
          Mux all downloaded streams to a video container (.mp4, .mkv, etc.) using ffmpeg. Note that
          existing files will be overwritten and downloaded streams will be deleted
      --prefer-audio-lang <PREFER_AUDIO_LANG>
          Preferred language when multiple audio streams with different languages are available.
          Must be in RFC 5646 format (eg. fr or en-AU). If a preference is not specified and
          multiple audio streams are present, the first one listed in the manifest will be
          downloaded
      --prefer-subs-lang <PREFER_SUBS_LANG>
          Preferred language when multiple subtitles streams with different languages are available.
          Must be in RFC 5646 format (eg. fr or en-AU). If a preference is not specified and
          multiple subtitles streams are present, the first one listed in the manifest will be
          downloaded
  -q, --quality <WIDTHxHEIGHT>
          Automatic selection of some standard resolution streams with highest bandwidth stream
          variant from playlist. possible values: [144p, 240p, 360p, 480p, 720p, hd, 1080p, fhd, 2k,
          1440p, qhd, 4k, 8k, highest, max, select-later] [default: select-later]
      --raw-prompts
          Raw style input prompts for old and unsupported terminals
      --retry-count <RETRY_COUNT>
          Maximum number of retries to download an individual segment [default: 15]
  -s, --skip
          Skip downloading and muxing alternative streams
  -t, --threads <THREADS>
          Maximum number of threads for parllel downloading of segments. Number of threads should be
          in range 1-16 (inclusive) [default: 5]
  -h, --help
          Print help information

Client Options:
      --cookies <COOKIES> <URL>
          Enable cookie store and fill it with some existing cookies. Example `--cookies "foo=bar;
          Domain=yolo.local" https://yolo.local`. This option can be used multiple times
      --enable-cookies
          Enable cookie store which allows cookies to be stored
      --header <KEY> <VALUE>
          Custom headers for requests. This option can be used multiple times
      --proxy-address <PROXY_ADDRESS>
          Set http or https proxy address for requests
      --user-agent <USER_AGENT>
          Update and set custom user agent for requests [default: "Mozilla/5.0 (Windows NT 10.0;
          Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.4951.64 Safari/537.36"]
```

## Alternatives

List of alternatives to vsd:

1. [N_m3u8DL-RE](https://github.com/nilaoda/N_m3u8DL-RE) is the best alternative to vsd. It also supports live playlist which vsd doesn't. It doesn't come with features like *capture* and *collect*. Also, CENC encrypted playlist decryption is slow.
2. [N_m3u8DL-CLI](https://github.com/nilaoda/N_m3u8DL-CLI) is good but it is not cross platform.
3. [m3u8-downloader](https://github.com/llychao/m3u8-downloader) is also good but it has very few customizable options.
4. [webvideo-downloader](https://github.com/jaysonlong/webvideo-downloader) opens websites using chrome and captures the m3u8 links and then downloads it. A similar functionality can achieved with vsd too by using *capture* and *collect* subcommands.
5. [dash-mpd-cli](https://github.com/emarsden/dash-mpd-cli) is very good for downloading DASH playlists. Also most of the vsd functionalities for parsing and downloading DASH playlists is taken for it's main project.

## License

Dual Licensed

- [Apache License, Version 2.0](https://www.apache.org/licenses/LICENSE-2.0) ([LICENSE-APACHE](LICENSE-APACHE))
- [MIT license](https://opensource.org/licenses/MIT) ([LICENSE-MIT](LICENSE-MIT))


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- `capture`
  - `--cookies`, `--directory`, `--save`, `--extensions` and `--resource-types` flags.
- `save`
  - Support for HLS `#EXT-X-MAP` tag.
  - Support for multi period DASH manifests.
  - Support for downloading playlists with single segment.
  - Support for parsing pssh box from initialization and displaying all key ids.
  - Support for browser cookies i.e. with `--cookies` flag.
  - Support for socks proxy.
  - `min` / `lowest` quality option for `--quality` flag.
  - `--skip-prompts`, `--all-keys` and `--no-decrypt` flags.

### Changed

- `capture`
  - `collect` sub-command is merged with `capture` sub-command.
  - `--build` flag is removed from `collect` sub-command.
- `collect` sub-command is removed.
- `decrypt` sub-command is removed.
- `extract`
  - `input` now only accepts single file.
  - `--format` flag is replaced with `--codec` flag.
- `merge`
  - `--ffmpeg` flag is replaced with `--type` flag.
- `save`
  - Video stream selection prompt is replaced with multi select prompt.
    Also, `--alternative` flag is removed and merged in this prompt. 
  - Use more accurate units (KiB, MiB, ..) to show download progress.
    Also, spinner is removed from progress bar.
  - `--cookies` flag is renamed as `--set-cookie`.
  - `--baseurl` flag is renamed as `--base-url`.
  - `--proxy-address` flag is renamed as `--proxy`.
  - `--quality` flag now also matches height if that specific resolution is not found.

### Fixed

- `capture`
  - Handle `CTRL+C` signal correctly.
- `save`
  - Some program panics when auto selecting streams using `--quality` flag.
  - `--directory` and `--header` flag implementation.
  - Unknown errors while extracting `stpp` and `application/ttml+xml` streams.
  - Use HLS `#EXT-X-KEY` tag more correctly.
  - DASH stream parsing logic.

## [0.2.5] - 2023-01-09

### Changed

- Do not use space character when saving file, instead use `vsd_*` prefix.
- `capture` and `collect` sub-commands are kept under optional cargo feature (`chrome`) but this feature is enabled by default.

### Fixed

- Relative url build using baseurl for local `.mpd` files.
- Segmentation fault when using threads more than 1.
- Subtitles saved as `.txt` but ffmpeg command uses `.vtt`.
- Match playlist kid(s) correctly with `--key` flag.

## [0.2.0] - 2022-10-08

### Added

- *DASH* support with decryption and subtitles.
- Sub-commands instead of a single command where *save* is the main sub-command.
- New singular progress bar for complete download progress.
- Better variant stream selection and display order.
- Improved support for playlists using byte range.
- Improved *capture* and *collect* sub-commands.
  - Using response received url when using *capture* sub-command.
  - Using chrome response for fetching playlists when using *collect* sub-command.

### Changed

- Default command is split into *save*, *capture* and *collect* sub-commands.
- Resume support is removed for now.

### Fixed

- `.vtt` -> `.srt` conversion ffmpeg command correction.
- No website scraping when extension is `.m3u`.

## [0.1.2] - 2022-07-09

### Added

- New `--build` flag.
- `.srt` subtitles collection with `--collect` flag.

### Fixed

- Not intercepting requests before navigating to website when using `--capture` and `--collect` flags.

## [0.1.0] - 2022-06-22

[Unreleased]: https://github.com/clitic/vsd/compare/v0.2.5...HEAD
[0.3.0]: https://github.com/clitic/vsd/compare/v0.2.5...v0.3.0
[0.2.5]: https://github.com/clitic/vsd/compare/v0.2.0...v0.2.5
[0.2.0]: https://github.com/clitic/vsd/compare/v0.1.2...v0.2.0
[0.1.2]: https://github.com/clitic/vsd/compare/v0.1.0...v0.1.2
[0.1.0]: https://github.com/clitic/vsd/releases/tag/v0.1.0


src/args.rs
use anyhow::{bail, Result};
use clap::{ArgGroup, Parser};
use kdam::term::Colorizer;
use reqwest::blocking::Client;
use reqwest::cookie::Jar;
use reqwest::header::{HeaderMap, HeaderName, HeaderValue};
use reqwest::{Proxy, Url};
use std::path::{Path, PathBuf};
use std::str::FromStr;
use std::sync::Arc;

#[allow(non_camel_case_types)]
#[derive(Debug, Clone)]
pub enum Quality {
    yt_144p,
    yt_240p,
    yt_360p,
    yt_480p,
    yt_720p,
    yt_1080p,
    yt_2k,
    yt_1440p,
    yt_4k,
    yt_8k,
    Resolution(u16, u16),
    Highest,
    SelectLater,
}

impl FromStr for Quality {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(match s.to_lowercase().as_str() {
            "144p" => Self::yt_144p,
            "240p" => Self::yt_240p,
            "360p" => Self::yt_360p,
            "480p" => Self::yt_480p,
            "720p" | "hd" => Self::yt_720p,
            "1080p" | "fhd" => Self::yt_1080p,
            "2k" => Self::yt_2k,
            "1440p" | "qhd" => Self::yt_1440p,
            "4k" => Self::yt_4k,
            "8k" => Self::yt_8k,
            "highest" | "max" => Self::Highest,
            "select-later" => Self::SelectLater,
            x if x.contains("x") => {
                if let (Some(w), Some(h)) = (x.split("x").nth(0), x.split("x").nth(1)) {
                    Self::Resolution(
                        w.parse::<u16>().map_err(|_| "invalid width".to_owned())?,
                        h.parse::<u16>().map_err(|_| "invalid height".to_owned())?,
                    )
                } else {
                    Err("incorrect resolution format".to_owned())?
                }
            }
            _ => Err(format!(
                "\npossible values: [{}]\nFor custom resolution use {}",
                [
                    "144p",
                    "240p",
                    "360p",
                    "480p",
                    "720p",
                    "hd",
                    "1080p",
                    "fhd",
                    "2k",
                    "1440p",
                    "qhd",
                    "4k",
                    "8k",
                    "highest",
                    "max",
                    "select-later",
                ]
                .iter()
                .map(|x| x.colorize("green"))
                .collect::<Vec<_>>().join(", "), "WIDTHxHEIGHT".colorize("green")
            ))?,
        })
    }
}

fn input_validator(s: &str) -> Result<(), String> {
    if s.starts_with("https://youtube.com")
        || s.starts_with("https://www.youtube.com")
        || s.starts_with("https://youtu.be")
    {
        Err("Youtube links aren't supported yet".to_owned())
    } else {
        Ok(())
    }
}

fn quality_validator(s: &str) -> Result<(), String> {
    let _ = s.parse::<Quality>()?;
    Ok(())
}

fn threads_validator(s: &str) -> Result<(), String> {
    let num_threads: usize = s.parse().map_err(|_| format!("`{}` isn't a number", s))?;
    if std::ops::RangeInclusive::new(1, 16).contains(&num_threads) {
        Ok(())
    } else {
        Err("number of threads should be in range `1-16`".to_string())
    }
}

fn proxy_address_validator(s: &str) -> Result<(), String> {
    if s.starts_with("http://") || s.starts_with("https://") {
        Ok(())
    } else {
        Err("Proxy address should start with `http://` or `https://` only".to_string())
    }
}

/// Command line program to download HLS video from websites and m3u8 links.
#[derive(Debug, Clone, Parser)]
#[clap(version, author = "clitic <clitic21@gmail.com>", about, group = ArgGroup::new("chrome").args(&["capture", "collect"]))]
pub struct Args {
    /// URL | .m3u8 | .m3u | .mpd | .xml
    #[clap(required = true, validator = input_validator)]
    pub input: String,

    /// Path of final downloaded video stream.
    /// For file extension any ffmpeg supported format could be provided.
    /// If playlist contains alternative streams vsd will try to transmux and trancode into single file using ffmpeg.
    #[clap(short, long)]
    pub output: Option<String>,

    /// Base url for all segments.
    /// Usually needed for local m3u8 file.
    #[clap(short, long)]
    pub baseurl: Option<String>,

    /// Automatic selection of some standard resolution streams with highest bandwidth stream variant from master playlist.
    /// possible values: [144p, 240p, 360p, 480p, 720p, hd, 1080p, fhd, 2k, 1440p, qhd, 4k, 8k, highest, max, select-later]
    #[clap(short, long, default_value = "select-later", value_name = "WIDTHxHEIGHT", validator = quality_validator)]
    pub quality: String,

    // /// Automatic selection of some standard resolution streams with highest bandwidth stream variant from master playlist.
    // #[clap(long, number_of_values = 2, value_names = &["width", "height"])]
    // pub resolution: Vec<u64>,
    /// Maximum number of threads for parllel downloading of segments.
    /// Number of threads should be in range 1-16 (inclusive).
    #[clap(short, long, default_value_t = 5, validator = threads_validator)]
    pub threads: u8,

    /// Maximum number of retries to download an individual segment.
    #[clap(long, default_value_t = 15)]
    pub retry_count: u8,

    /// Raw style input prompts for old and unsupported terminals.
    #[clap(long)]
    pub raw_prompts: bool,

    /// Resume a download session.
    /// Download session can only be resumed if download session json file is present.
    #[clap(short, long)]
    pub resume: bool,

    /// Download alternative streams such as audio and subtitles streams from master playlist instead of variant video streams.
    #[clap(short, long)]
    pub alternative: bool,

    /// Skip downloading and muxing alternative streams.
    #[clap(short, long)]
    pub skip: bool,

    /// TODO: Decryption keys.
    /// This option can be used multiple times.
    #[clap(short, long, multiple_occurrences = true, value_name = "<KID:KEY>|KEY")]
    pub key: Vec<String>,

    /// TODO: Record duration for live playlist in seconds.
    #[clap(long)]
    pub record_duration: Option<f32>,

    /// TODO: Directory path
    #[clap(long)]
    pub save_directory: Option<String>,

    /// Launch Google Chrome to capture requests made to fetch .m3u8 (HLS) and .mpd (Dash) files.
    #[clap(long, help_heading = "CHROME OPTIONS")]
    pub capture: bool,

    /// Launch Google Chrome and collect .m3u8 (HLS), .mpd (Dash) and subtitles from a website and save them locally.
    #[clap(long, help_heading = "CHROME OPTIONS")]
    pub collect: bool,

    /// Launch Google Chrome without a window for interaction.
    /// This option should must be used with `--capture` or `--collect` flag only.
    #[clap(long, requires = "chrome", help_heading = "CHROME OPTIONS")]
    pub headless: bool,

    /// Build http links for all uri present in .m3u8 file while collecting it.
    /// Resultant .m3u8 file can be played and downloaded directly without the need of `--baseurl` flag.
    /// This option should must be used with `--collect` flag only.
    #[clap(long, requires = "collect", help_heading = "CHROME OPTIONS")]
    pub build: bool,

    /// Custom headers for requests.
    /// This option can be used multiple times.
    #[clap(long, multiple_occurrences = true, number_of_values = 2, value_names = &["KEY", "VALUE"], help_heading = "CLIENT OPTIONS")]
    pub header: Vec<String>, // Vec<Vec<String>> not supported

    /// Update and set custom user agent for requests.
    #[clap(
        long,
        default_value = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.4951.64 Safari/537.36",
        help_heading = "CLIENT OPTIONS"
    )]
    pub user_agent: String,

    /// Set http or https proxy address for requests.
    #[clap(long, validator = proxy_address_validator, help_heading = "CLIENT OPTIONS")]
    pub proxy_address: Option<String>,

    /// Enable cookie store which allows cookies to be stored.
    #[clap(long, help_heading = "CLIENT OPTIONS")]
    pub enable_cookies: bool,

    /// Enable cookie store and fill it with some existing cookies.
    /// Example `--cookies "foo=bar; Domain=yolo.local" https://yolo.local`.
    /// This option can be used multiple times.
    #[clap(long, multiple_occurrences = true, number_of_values = 2, value_names = &["COOKIES", "URL"], help_heading = "CLIENT OPTIONS")]
    pub cookies: Vec<String>, // Vec<Vec<String>> not supported
}

impl Args {
    pub fn client(&self) -> Result<Arc<Client>> {
        let mut client_builder = Client::builder().user_agent(&self.user_agent);

        if !self.header.is_empty() {
            let mut headers = HeaderMap::new();

            for i in (0..headers.len()).step_by(2) {
                headers.insert(
                    self.header[i].parse::<HeaderName>()?,
                    self.header[i + 1].parse::<HeaderValue>()?,
                );
            }

            client_builder = client_builder.default_headers(headers);
        }

        if let Some(proxy) = &self.proxy_address {
            if proxy.starts_with("https") {
                client_builder = client_builder.proxy(Proxy::https(proxy)?);
            } else if proxy.starts_with("http") {
                client_builder = client_builder.proxy(Proxy::http(proxy)?);
            }
        }

        if self.enable_cookies || !self.cookies.is_empty() {
            client_builder = client_builder.cookie_store(true);
        }

        if !self.cookies.is_empty() {
            let jar = Jar::default();

            for i in (0..self.cookies.len()).step_by(2) {
                jar.add_cookie_str(&self.cookies[i], &self.cookies[i + 1].parse::<Url>()?);
            }

            client_builder = client_builder.cookie_provider(Arc::new(jar));
        }

        Ok(Arc::new(client_builder.build()?))
    }

    pub fn get_url(&self, uri: &str) -> Result<String> {
        if uri.starts_with("http") {
            Ok(uri.to_owned())
        } else if let Some(baseurl) = &self.baseurl {
            Ok(Url::parse(baseurl)?.join(uri)?.to_string())
        } else {
            if !self.input.starts_with("http") {
                bail!(
                    "Non HTTP input should have {} set explicitly.",
                    "--baseurl".colorize("bold green")
                )
            }

            Ok(Url::parse(&self.input)?.join(uri)?.to_string())
        }
    }

    pub fn tempfile(&self) -> String {
        let output = self
            .input
            .split('?')
            .next()
            .unwrap()
            .split('/')
            .last()
            .unwrap();

        let output = if output.ends_with(".m3u") || output.ends_with(".m3u8") {
            if output.ends_with(".ts.m3u8") {
                output.trim_end_matches(".m3u8").to_owned()
            } else {
                let mut path = PathBuf::from(&output);
                path.set_extension("ts");
                path.to_str().unwrap().to_owned()
            }
        } else if output.ends_with(".mpd") || output.ends_with(".xml") {
            let mut path = PathBuf::from(&output);
            path.set_extension("m4s");
            path.to_str().unwrap().to_owned()
        } else {
            let mut path = PathBuf::from(
                output
                    .replace('<', "-")
                    .replace('>', "-")
                    .replace(':', "-")
                    .replace('\"', "-")
                    .replace('/', "-")
                    .replace('\\', "-")
                    .replace('|', "-")
                    .replace('?', "-"),
            );
            path.set_extension("mp4");
            path.to_str().unwrap().to_owned()
        };

        if Path::new(&output).exists() && !self.resume {
            let stemed_path = Path::new(&output).file_stem().unwrap().to_str().unwrap();
            let ext = Path::new(&output).extension().unwrap().to_str().unwrap();

            for i in 1.. {
                let core_file_copy = format!("{} ({}).{}", stemed_path, i, ext);

                if !Path::new(&core_file_copy).exists() {
                    return core_file_copy;
                }
            }
        }

        output
    }

    pub fn input_type(&self) -> InputType {
        let url = self.input.split('?').next().unwrap();

        if url.starts_with("http") {
            if url.ends_with(".m3u") || url.ends_with(".m3u8") || url.ends_with("ts.m3u8") {
                InputType::HlsUrl
            } else if url.ends_with(".mpd") || url.ends_with(".xml") {
                InputType::DashUrl
            } else {
                InputType::Website
            }
        } else {
            if url.ends_with(".m3u") || url.ends_with(".m3u8") || url.ends_with("ts.m3u8") {
                InputType::HlsLocalFile
            } else if url.ends_with(".mpd") || url.ends_with(".xml") {
                InputType::DashLocalFile
            } else {
                InputType::LocalFile
            }
        }
    }
}
pub enum InputType {
    HlsUrl,
    DashUrl,
    Website,
    HlsLocalFile,
    DashLocalFile,
    LocalFile,
}

impl InputType {
    pub fn is_website(&self) -> bool {
        match &self {
            Self::Website => true,
            _ => false,
        }
    }

    pub fn is_hls(&self) -> bool {
        match &self {
            Self::HlsUrl | Self::HlsLocalFile => true,
            _ => false,
        }
    }

    pub fn is_dash(&self) -> bool {
        match &self {
            Self::DashUrl | Self::DashLocalFile => true,
            _ => false,
        }
    }
}
